[SPARK-25353][SQL] executeTake in SparkPlan is modified to avoid unnecessary decoding.

Dooyoung-Hwang opened the pull request:
## What changes were proposed in this pull request?
In some cases, executeTake in SparkPlan could decode more than necessary.

For example, in case of below odd/even number partitioning, total row's count from partitions will be 100, although it is limited with 51. And 'executeTake' in SparkPlan decodes all of them, "49" rows of which are unnecessarily decoded.

```scala
spark.sparkContext.parallelize((0 until 100).map(i => (i, 1))).toDF()
      .repartitionByRange(2, $"_1" % 2).limit(51).collect()
```

By using a iterator of the scalar collection, we can make ensure that at most n rows are decoded.

## How was this patch tested?
Existing unit tests that call limit function of DataFrame.

testOnly *SQLQuerySuite
testOnly *DataFrameSuite

Commit:
Refactoring executeTake(n: Int) in SparkPlan

Using view of scala collections, deserialize rows at most n times.

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

Dooyoung-Hwang commented:
cc: @maropu @viirya @kiszk @HyukjinKwon 
This PR is separated from https://github.com/apache/spark/pull/22219 by @maropu's opinion.

AmplabJenkins commented:
Can one of the admins verify this patch?

kiszk commented:
Thank you for your update.
Is it better to add test case to confirm the state of the internal structures is as you expected? @maropu

Dooyoung-Hwang commented:
@kiszk 
It is impossible counting decoded rows without modify SparkPlan, because there is no way of counting iterated size.

Instead I can simulate this patch in Scala WorkSheet with below code.

```scala
var decodeCount = 0
def decoding(buf: Array[Int]): Iterator[String] = {
  new Iterator[String] {
    var remain = buf.sum
    var index = 0
    override def hasNext: Boolean = remain > 0
    override def next(): String = {
      while (buf(index) == 0) index += 1
      buf(index) -= 1
      remain -= 1
      decodeCount += 1    // increase decodeCount
      f"[decode Result:$remain]"
    }
  }
}

// reset decodeCount
decodeCount = 0

// Before Patch : decode without scala view
val buf = new ArrayBuffer[String]
val inputIter = Array(Array(2, 2, 2), Array(2), Array(2)).iterator
while (inputIter.hasNext) buf ++= Array(inputIter.next()).flatMap(decoding)
val result1 = buf.take(3).toArray

// ensure decode count is 10
assert(decodeCount == 10)

// reset decodeCount
decodeCount = 0

// After Patch : decode with scala view
val result2 = ArrayBuffer(Array(2, 2, 2), Array(2), Array(2)).toArray.view
  .flatMap(decoding).take(3).force

// ensure decode count is 3
assert(decodeCount == 3)

// assert same element
assert(result1 sameElements result2)
```

maropu commented:
btw, do we have any actual performance benefit (wall time) from this patch?

HyukjinKwon commented:
Let me leave this ok to test since there looks a progress here anyway.

HyukjinKwon commented:
ok to test

SparkQA commented:
**[Test build #95859 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/95859/testReport)** for PR 22347 at commit [`a8f1481`](https://github.com/apache/spark/commit/a8f14817ce3f52f710c3341148c2e1f3374335eb).

SparkQA commented:
**[Test build #95859 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/95859/testReport)** for PR 22347 at commit [`a8f1481`](https://github.com/apache/spark/commit/a8f14817ce3f52f710c3341148c2e1f3374335eb).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/95859/
Test FAILed.

Dooyoung-Hwang commented:
I tested in my local PC. 3.3 GHz Intel Core i5, and selected 400,000 rows x 25 times.
I took a total execution time between decodeUnsafeRows.
My tested data is skewed, so gathered rows from executor are distributed between 400000 and  800000. 
The average execution time decreased from 175.92ms to 93.52ms.
Memory usage also improved, and total GC Time is decreased from 13.883 sec to 10.764 sec.

## Before Patch
### GC statics

S0 | S1 | E | O | M | CCS | YGC | YGCT | FGC | FGCT | GCT
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
0 | 100 | 24.92 | 48.66 | 96.92 | 88.25 | 150 | 13.883 | 0 | 0 | 13.883

### Wall time : AVG 175.92 ms
Row Count | Decode Time(ms)
-- | --
428942 | 73
473726 | 106
476322 | 78
509996 | 83
510590 | 124
556896 | 94
556896 | 362
595272 | 193
595272 | 175
642478 | 120
644970 | 279
679544 | 269
693354 | 116
723532 | 124
729912 | 136
730218 | 120
730246 | 184
773640 | 183
774148 | 380
810198 | 128
811606 | 131
859090 | 138
895474 | 314
895954 | 339
939636 | 149

## After Patch : 93.52ms

### GC statics
S0 | S1 | E | O | M | CCS | YGC | YGCT | FGC | FGCT | GCT
-- | -- | -- | -- | -- | -- | -- | -- | -- | -- | --
0 | 100 | 81.37 | 33.34 | 97.21 | 88.35 | 127 | 10.764 | 0 | 0 | 10.764

### Wall time : AVG 93.52 ms
Row Count | Decode time (ms)
-- | --
421922 | 61
422516 | 180
422850 | 110
473218 | 62
473218 | 103
473438 | 115
507198 | 60
554606 | 144
557202 | 119
601392 | 71
642652 | 61
645276 | 64
679036 | 64
679036 | 63
729624 | 242
729652 | 62
729912 | 131
773814 | 122
774234 | 62
807908 | 59
810198 | 64
814900 | 72
844772 | 59
858582 | 127
858582 | 61

Dooyoung-Hwang commented:
Jenkins, retest this please

SparkQA commented:
**[Test build #95892 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/95892/testReport)** for PR 22347 at commit [`a8f1481`](https://github.com/apache/spark/commit/a8f14817ce3f52f710c3341148c2e1f3374335eb).

SparkQA commented:
**[Test build #95892 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/95892/testReport)** for PR 22347 at commit [`a8f1481`](https://github.com/apache/spark/commit/a8f14817ce3f52f710c3341148c2e1f3374335eb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/95892/
Test PASSed.

dongjoon-hyun commented:
Retest this please.

dongjoon-hyun commented:
Thank you for your first contribution, @Dooyoung-Hwang ! So, this is a spin-off PR from STS?

Dooyoung-Hwang commented:
Thank you for review. 
Yes, ThriftServer will use intermediate "collection view" in this PR.
And [Original PR of ThriftServer](https://github.com/apache/spark/pull/22219) will be updated accordingly ,if this PR is merged.

SparkQA commented:
**[Test build #96870 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96870/testReport)** for PR 22347 at commit [`a8f1481`](https://github.com/apache/spark/commit/a8f14817ce3f52f710c3341148c2e1f3374335eb).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/96870/
Test PASSed.

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala:
```diff
@@ -348,30 +349,30 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
-    if (buf.size > n) {
-      buf.take(n).toArray
+    if (scannedRowCount > n) {
+      buf.toArray.view.flatMap(decodeUnsafeRows).take(n).force
```
Can we simplify like the following?
```scala
      buf.flatMap(decodeUnsafeRows).take(n).toArray
```

Dooyoung-Hwang replied:
If the part of calling "view" is removed, no behavior changes compared to before.  If you don't want to use view here, what about "iterator" instead of "view".
It would be like this.

buf.iterator.flatMap(decodeUnsafeRows).take(n).toArray

dongjoon-hyun commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala:
```diff
@@ -348,30 +349,30 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
+      buf.toArray.view.flatMap(decodeUnsafeRows).take(n).force
     } else {
-      buf.toArray
+      buf.toArray.view.flatMap(decodeUnsafeRows).force
```
`buf.flatMap(decodeUnsafeRows).toArray` ?

Dooyoung-Hwang replied:
Ok~

Commit:
Change view to iterator.

SparkQA commented:
**[Test build #96937 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96937/testReport)** for PR 22347 at commit [`8666272`](https://github.com/apache/spark/commit/86662722e53bfcae2c75e61d170c983abd599b3a).

Dooyoung-Hwang commented:
I added example code of issue case to the content of PR.

SparkQA commented:
**[Test build #96937 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/96937/testReport)** for PR 22347 at commit [`8666272`](https://github.com/apache/spark/commit/86662722e53bfcae2c75e61d170c983abd599b3a).
 * This patch **fails SparkR unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/96937/
Test FAILed.

tooptoop4 commented:
@dongjoon-hyun can this be merged?

kiszk commented:
retest this please

SparkQA commented:
**[Test build #97242 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97242/testReport)** for PR 22347 at commit [`8666272`](https://github.com/apache/spark/commit/86662722e53bfcae2c75e61d170c983abd599b3a).

SparkQA commented:
**[Test build #97242 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/97242/testReport)** for PR 22347 at commit [`8666272`](https://github.com/apache/spark/commit/86662722e53bfcae2c75e61d170c983abd599b3a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/97242/
Test PASSed.

tooptoop4 commented:
please merge

AmplabJenkins commented:
Can one of the admins verify this patch?

kiszk commented:
Is it better to revisit this? cc @maropu

maropu commented:
Looks ok to me, but I think we need more comments from other reviewers... cc: @dongjoon-hyun @viirya @cloud-fan

maropu commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala:
```diff
@@ -348,30 +349,30 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
-    if (buf.size > n) {
-      buf.take(n).toArray
+    if (scannedRowCount > n) {
+      buf.iterator.flatMap(decodeUnsafeRows).take(n).toArray
```
nit: can you put some comments about why we need `.iterator` here? we need it to avoid unnecessary decoding, right?

cloud-fan replied:
nit: since this is perf critical code path, I think we can optimize it further, since we know the length of the result array.
```
val result = new Array[InternalRow](n)
while (result.length < n) {
  // decode
}
result
```

cloud-fan commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12254/
Test PASSed.

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala:
```diff
@@ -348,30 +349,30 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
+      buf.iterator.flatMap(decodeUnsafeRows).take(n).toArray
     } else {
-      buf.toArray
+      buf.flatMap(decodeUnsafeRows).toArray
```
ditto

SparkQA commented:
**[Test build #107061 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107061/testReport)** for PR 22347 at commit [`8666272`](https://github.com/apache/spark/commit/86662722e53bfcae2c75e61d170c983abd599b3a).

cloud-fan commented:
LGTM, @maropu can you take it over if @Dooyoung-Hwang is not active?

maropu commented:
sure.

@Dooyoung-Hwang are you there?

SparkQA commented:
**[Test build #107061 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107061/testReport)** for PR 22347 at commit [`8666272`](https://github.com/apache/spark/commit/86662722e53bfcae2c75e61d170c983abd599b3a).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107061/
Test PASSed.

Dooyoung-Hwang commented:
@maropu
Thank you for review. I will prepare for a patch following @cloud-fan's comment.

Commit:
executeTask write decoded-rows to fixed size array.

SparkQA commented:
**[Test build #107078 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107078/testReport)** for PR 22347 at commit [`e09a10d`](https://github.com/apache/spark/commit/e09a10dac08f2fe2627e1c96b49de379e86ba3cf).

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala:
```diff
@@ -438,16 +439,27 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
 
-      buf ++= res.flatMap(decodeUnsafeRows)
-
+      val partResIterator = res.iterator
```
a bit better for perf:
```
var i = 0
while (scannedRowCount < n && i < res.length) {
  val rows = decodeUnsafeRows(res(i))
  while (scannedRowCount < n && rows.hasNext) {
    buf(scannedRowCount) = rows.next()
    scannedRowCount += 1
  }
  i += 1
}
```

It uses array indexing instead of iterator, and it inlines the method.

Commit:
1. Move writeArray to inline.
2. Change the type of scannedRowCount to Int, because it is always smaller than n.

SparkQA commented:
**[Test build #107082 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107082/testReport)** for PR 22347 at commit [`8685e2c`](https://github.com/apache/spark/commit/8685e2c28c8020f558d59124b470a54448de3a71).

SparkQA commented:
**[Test build #107078 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107078/testReport)** for PR 22347 at commit [`e09a10d`](https://github.com/apache/spark/commit/e09a10dac08f2fe2627e1c96b49de379e86ba3cf).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107078/
Test FAILed.

SparkQA commented:
**[Test build #107082 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107082/testReport)** for PR 22347 at commit [`8685e2c`](https://github.com/apache/spark/commit/8685e2c28c8020f558d59124b470a54448de3a71).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107082/
Test FAILed.

maropu commented:
retest this please

SparkQA commented:
**[Test build #107091 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107091/testReport)** for PR 22347 at commit [`8685e2c`](https://github.com/apache/spark/commit/8685e2c28c8020f558d59124b470a54448de3a71).

maropu commented:
LGTM, pending Jenkins.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12270/
Test PASSed.

SparkQA commented:
**[Test build #107091 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107091/testReport)** for PR 22347 at commit [`8685e2c`](https://github.com/apache/spark/commit/8685e2c28c8020f558d59124b470a54448de3a71).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107091/
Test FAILed.

maropu commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12275/
Test PASSed.

SparkQA commented:
**[Test build #107095 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107095/testReport)** for PR 22347 at commit [`8685e2c`](https://github.com/apache/spark/commit/8685e2c28c8020f558d59124b470a54448de3a71).

Dooyoung-Hwang commented:
Maybe this test code in DataFrameSuite causes OOM.

```
   // SPARK-12340: overstep the bounds of Int in SparkPlan.executeTake
    checkAnswer(
      spark.range(2).toDF().limit(2147483638),
      Row(0) :: Row(1) :: Nil
    )
```

maropu commented:
oh.... this pr makes the memory pressure higher in the query?

Commit:
Fix OOM in test suite.

- Incrementally increase the size of output array.

SparkQA commented:
**[Test build #107098 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107098/testReport)** for PR 22347 at commit [`7266236`](https://github.com/apache/spark/commit/72662360ceb7ee9fc0aa036a7c1d3e05140297c8).

cloud-fan commented:
Ah I see the problem, we should not create the result array ahead, as the limit can be large but the data only has very few records.

maropu commented:
is that bad like this? I like the previous version because that was simple:
```
val rowCount = if (scannedRowCount < n) scannedRowCount else n
val buf = new Array[InternalRow](rowCount)
```

cloud-fan commented on sql/core/src/main/scala/org/apache/spark/sql/execution/SparkPlan.scala:
```diff
@@ -423,31 +424,37 @@ abstract class SparkPlan extends QueryPlan[SparkPlan] with Logging with Serializ
+
+      var i = 0
+      while (scannedRowCount < n && i < res.length) {
+        val rows = decodeUnsafeRows(res(i)._2)
```
nit:
```
val rows = decodeUnsafeRows(res(i)._2)
val rowsToTake = if (n - scannedRowCount >= res(i)._1) {
  rows.toArray
} else {
  rows.take(n - scannedRowCount).toArray
}
scannedRowCount += rowsToTake.length
buf ++= rowsToTake
```

BTW do we still need `scannedRowCount`? It's always same as `buf.length`

Dooyoung-Hwang replied:
Thank you for review. Looks much better!

Dooyoung-Hwang replied:
Yes, we can remove scannedRowCount.

SparkQA commented:
**[Test build #107095 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107095/testReport)** for PR 22347 at commit [`8685e2c`](https://github.com/apache/spark/commit/8685e2c28c8020f558d59124b470a54448de3a71).
 * This patch **fails Spark unit tests**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107095/
Test FAILed.

Commit:
Nit : Refactor decoding block.

SparkQA commented:
**[Test build #107103 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107103/testReport)** for PR 22347 at commit [`29e27aa`](https://github.com/apache/spark/commit/29e27aa607d0b03e05560d1e18617fe1a0403a98).

SparkQA commented:
**[Test build #107103 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107103/testReport)** for PR 22347 at commit [`29e27aa`](https://github.com/apache/spark/commit/29e27aa607d0b03e05560d1e18617fe1a0403a98).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

SparkQA commented:
**[Test build #107098 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107098/testReport)** for PR 22347 at commit [`7266236`](https://github.com/apache/spark/commit/72662360ceb7ee9fc0aa036a7c1d3e05140297c8).
 * This patch **fails due to an unknown error code, -9**.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107103/
Test FAILed.

AmplabJenkins commented:
Merged build finished. Test FAILed.

AmplabJenkins commented:
Test FAILed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107098/
Test FAILed.

maropu commented:
retest this please

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/testing-k8s-prb-make-spark-distribution-unified/12286/
Test PASSed.

SparkQA commented:
**[Test build #107107 has started](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107107/testReport)** for PR 22347 at commit [`29e27aa`](https://github.com/apache/spark/commit/29e27aa607d0b03e05560d1e18617fe1a0403a98).

SparkQA commented:
**[Test build #107107 has finished](https://amplab.cs.berkeley.edu/jenkins/job/SparkPullRequestBuilder/107107/testReport)** for PR 22347 at commit [`29e27aa`](https://github.com/apache/spark/commit/29e27aa607d0b03e05560d1e18617fe1a0403a98).
 * This patch passes all tests.
 * This patch merges cleanly.
 * This patch adds no public classes.

AmplabJenkins commented:
Merged build finished. Test PASSed.

AmplabJenkins commented:
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
https://amplab.cs.berkeley.edu/jenkins//job/SparkPullRequestBuilder/107107/
Test PASSed.

cloud-fan commented:
thanks, merging to master!
